### ISSUE TITLE
Why the ratio in PPOTrainer use the same policy model rather than old and new policy model to generate?

### ISSUE THREAD

**Issue by Lynnzake** [huggingface/trl#3084]
ratio in PPO:
$$
\frac{\pi_{\theta}}{\pi_{\theta_{old}}}
$$
But The implementation in PPOTrainer, seems like it use the same policy model to generate the log\_probs and new\_log\_probs
```python
with unwrap_model_for_generation(
      self.model, self.accelerator, gather_deepspeed3_params=self.args.ds3_gather_for_generation
  ) as unwrapped_model:
      query_responses, logitss = batch_generation(
          unwrapped_model.policy,
          queries,
          args.local_rollout_forward_batch_size,
          processing_class.pad_token_id,
          generation_config,
      )

  for i in range(0, queries.shape[0], args.local_rollout_forward_batch_size):
      query = queries[i : i + args.local_rollout_forward_batch_size]
      query_response = query_responses[i : i + args.local_rollout_forward_batch_size]
      response = query_response[:, context_length:]
      logits = logitss[i : i + args.local_rollout_forward_batch_size]
      logprob = selective_log_softmax(logits, response)
```
**new\_log\_probs**: 
```python
for ppo_epoch_idx in range(args.num_ppo_epochs):
                b_inds = np.random.permutation(args.local_batch_size)
                minibatch_idx = 0
                for mini_batch_start in range(0, args.local_batch_size, args.local_mini_batch_size):
                    mini_batch_end = mini_batch_start + args.local_mini_batch_size
                    mini_batch_inds = b_inds[mini_batch_start:mini_batch_end]
                    gradient_accumulation_idx = 0
                    for micro_batch_start in range(0, args.local_mini_batch_size, args.per_device_train_batch_size):
                        with accelerator.accumulate(model):
                            micro_batch_end = micro_batch_start + args.per_device_train_batch_size
                            micro_batch_inds = mini_batch_inds[micro_batch_start:micro_batch_end]
                            mb_advantage = advantages[micro_batch_inds]
                            mb_responses = responses[micro_batch_inds]
                            mb_query_responses = query_responses[micro_batch_inds]
                            mb_logprobs = logprobs[micro_batch_inds]
                            mb_return = returns[micro_batch_inds]
                            mb_values = values[micro_batch_inds]

                            output, vpred_temp = forward(model, mb_query_responses, processing_class.pad_token_id)
                            logits = output.logits[:, context_length - 1 : -1]
                            logits /= args.temperature + 1e-7
                            new_logprobs = selective_log_softmax(logits, mb_responses)
                            new_logprobs = torch.masked_fill(
                                new_logprobs, padding_mask[micro_batch_inds], INVALID_LOGPROB
                            )
```
```python
logprobs_diff = new_logprobs - mb_logprobs
ratio = torch.exp(logprobs_diff)
```
And the model update it's param after calculate the ratio.
So there is no old probs generated by old policy model? Why?

**Comment by qgallouedec** [comment]
Hey @Lynnzake, sorry for the very slow reply, we're catching up on the backlog.

This is working as intended: PPO doesn't need a separate "old policy" *model*, only the old policy's *log-probs*. Those are computed once per batch at rollout/generation time, before any gradient step, and cached in `logprobs` see https://github.com/huggingface/trl/blob/378fe7c04fb2f60d42b99803ab99de4c43653ef3/trl/experimental/ppo/ppo_trainer.py#L700-L703
During the update phase, `new_logprobs` are recomputed each micro-batch with the current (updating) policy, and `ratio = exp(new_logprobs - old_logprobs)` https://github.com/huggingface/trl/blob/378fe7c04fb2f60d42b99803ab99de4c43653ef3/trl/experimental/ppo/ppo_trainer.py#L820-L842
Because `num_ppo_epochs` and the minibatch loop take several gradient steps on the same rollout, the policy drifts from that snapshot and the ratio becomes meaningful, on the very first step it's ~1, which is exactly what PPO expects.
